### ISSUE TITLE
[RFC] Use TypeScript strict compiler option?

### ISSUE THREAD

**Issue by timfish** [aurelia/aurelia#254]
# 💬 RFC

Firstly I'd like to say 👍 for choosing TypeScript. It makes it much easier to contribute without requiring intimate knowledge of the entire codebase. 

Admittedly, we haven't gone back over large legacy TypeScript projects and enabled strict mode, but when we start a new project, or refactor into modules, we now always default to enabling the TypeScript strict compile option.

This is a shortcut for enabling: 
```
noImplicitAny
noImplicitThis 
alwaysStrict 
strictNullChecks
strictFunctionTypes 
strictPropertyInitialization
```

We've found it a great way to improve code quality as `strictNullChecks` specifically forces developers to consider all the places where something might be undefined. 

Maybe you've already considered this and decided that the performance overhead of extra null checks is not worth it? You almost certainly wouldn't want it enabled for tests.

If strict mode is too cumbersome for Aurelia core, its still worth using 'nullable' type annotations for public API's (ie. `ISomeElement | undefined`) so that consumers of Aurelia using strict mode know exactly what return values they have to cater for.

Once the vNext changes have settled down, I can devote some time to work on this, I just didn't want to do that if you've already discounted it usage!

**Comment by fkleuver** [comment]
Thank you for opening this issue, I think it's good to have a dedicated discussion about the typings.

I think I speak for most of the core team (myself included) when saying: the stricter, the better.

Not only does it help find and prevent bugs, it's also a form of documentation. I for one am a big advocate of self-documenting code: if you can explain logic by picking the right variable names and organizing the code in a certain manner, then that's better than by doing it via comments.

The drawback of TS/JS is that, unlike with a compiled language such as C#, more verbose ways of writing something often leads to worse performance. The most performant code in JavaScript also tends to be the least readable code. So wherever we can improve that situation via stricter typings (and JS doc comments for intellisense), we definitely should.

> Maybe you've already considered this and decided that the performance overhead of extra null checks is not worth it?

We actually left it out because the code was changing so frequently and in such large volumes that strict mode would create too much overhead with refactoring. We needed to iterate fast without having to change along a ton of typings all the time.

In the mean time I've been relying on unit and integration tests to pound the code with as many combinations of nonsense as possible, in order to discover potential edge cases and patch them up where needed.

All that said, I think now is actually an okay time to start looking at strictness of typings in the less volatile areas of the code.

I don't think the performance overhead of extra null checks would really be an issue except for maybe in the hottest code paths.
Another thing I'd love to see is more strict enforcing of pre- and post conditions and invariants. "code contracts" of sorts. And have them clearly documented at the API boundaries, but also internally. For the internal cases this means we will be able to make certain assumptions in certain places and we can simply use `!` to assert when we know something is not null.

**Comment by fkleuver** [comment]
I'd like to start with enabling `strictPropertyInitialization` as soon as possible. Here's why:
- We *need* to initialize most properties eagerly anyway, typically for performance-critical code that uses explicit null checks (performs better than falsey checks). We're already doing this for the most part, and wherever we're *not* doing it, the code tends to be buggy.
- Eagerly initializing all properties in a particular order also reduces the number of hidden classes that need to be generated by the browser
- Forcing all properties to be defined in the constructor makes the code easier to debug. Otherwise it first loops through the constructor and then goes up to the properties, it's messy.

After I'm done refactoring the lifecycles and the PR is merged, I'd like to enable this option right away and start with that. It should be a pretty quick and straight-forward exercise of moving property assignments down or adding it where it's absent. What do you think @EisenbergEffect  @bigopon

**Comment by EisenbergEffect** [comment]
100% agreement from me.

**Comment by timfish** [comment]
Sorry that after suggesting this, I haven't found time to help out as the end of 2018 was pretty crazy for me. I did find some time this morning to pull the latest code and have another look around and its looking very good!

I did notice in `tslint.json`:
`"no-null-keyword": false, // use undefined to mean not initialized and null to mean without a value`

I can see that it allows you to store that extra state without requiring extra memory, but are there many places where the distinction between the two matters enough to have the extra complication? It means you end up with code like [in this (admittedly abandoned) PR](https://github.com/aurelia/aurelia/pull/363/files#diff-10e601058a29913ecb0c5f90ea774b93R323).

The [coding guidelines](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined) for the TypeScript compiler specifically disallow the use of `null` but they're obviously not dealing with the DOM or other APIs which return `null`.

**Comment by fkleuver** [comment]
For completeness sake, here is part of my reasoning for `null` (I may have forgotten some motivations that I changed my mind about over time):
- Performance: eagerly initializing properties causes hot paths to be found sooner by the browser and gives it fewer variations of hidden classes to generate.
- Early errors / validation: an explicit `=== null` check on a type that is (partially) unknown or cast from another type, when it evaluates to `false`, invalidates any assumption that may have been made about its shape (we are dealing with user input and this is a kind of "duck-typing")
- Readability: it is a shorter keyword than `undefined`
- Familiarity: `null` is more well-known from other languages, in particular the .NET crowd that we tend to appeal to

I want to stick to explicit `===` checks and avoid truthy/falsey value coercion as much as possible. This is for performance (on hot paths it can make a huge difference) as well as readability (having to look around wondering whether something could also be zero or an empty string, as opposed to just `null`, is not very productive).

> It means you end up with code like in this (admittedly abandoned) PR.

This is precisely why I went all-in on having *everything* eagerly and explicitly initialized to `null`, so that the `=== null || === undefined` wouldn't be necessary.

Recently I've been looking into the internals of the TypeScript compiler (as well as their tslint/tsconfig rules) and, combined with the experience of working on vNext for this past half year, I came to the conclusion that working with `null` in this manner possibly introduces more problems than it solves.

The primary thing is that several folks have questioned it, meaning it's different from what they're used to and they feel they need to adapt. This makes the framework less contributor friendly and stands in direct conflict with our goal to make it much *more* contributor friendly (we need contributors, after all).

For this reason I can get behind dropping `null` in favor of `undefined`.

@bigopon @EisenbergEffect thoughts?
